ctsit/nacculator
# Add ability for NACCulator to process date and rater header info on individual forms

This change adds logic to NACCulator's builder.py files that checks each individual form for date and rater information before it checks the main header form. This allows centers to specify which doctor filled out a form, and if they were filled on different days. If a center has not added these fields to their project, then NACCulator's functionality has not changed other than a few bug fixes.

Small fixes include: 
The C2T form for telephone followup packets is now optional. The program will no longer assume the C2T is present.
The LBD event detection is improved (it will search for more variations of the B1L form's fieldname).
"NOGDS" in form B6 has been added to set_to_zero_if_blank to account for the TFP forms.

**Why was this change made?** _Including an issue number is sufficient. Otherwise, briefly explain the benefit of the change._

This change was made to address Elijah Rockers' request to add custom form header functionality to NACCulator. This pull request uses the same builder.py logic as in https://github.com/ctsit/nacculator/pull/118 and I made sure to fix the same typo that Eli found in the uds3 fvp forms.py file.
My code differs by adding a few other bug fixes and leaving the forms.py files otherwise unchanged, so that the unit tests still pass.


## Verification

1. Collect a data export from your center's REDCap project.
2. From the command-line, run:
`redcap2nacc -ivp -lbdsv <input_data.csv >output_data.txt`
(or a command with a similar flag; I am using lbd short version to demonstrate the improved event detection)
3. The output should look like: Regular NACCulator output, without any KeyErrors involved. If your center has custom form header info, then each individual form should have that custom date and rater info in the header of each line.
4. If you pass _some bad value_, the output should look like: Regular NACCulator error messages, with no false positives.

- Verify the thing does this: Handles custom form header info, and correctly processes TFP v3.2 data.
- Verify the thing does not do that: Skips forms that are required, nor includes optional forms that are empty.


## Affirmations

All of these should have a check by them. Any exception requires an explanation.
